pkg/distributor/push_test.go

```diff
@@ -1293,6 +1293,191 @@ func TestHandler_ToHTTPStatus(t *testing.T) {
+		originalMsg = "this is an error"
+	)
+	originalErr := errors.New(originalMsg)
+	replicasNotMatchErr := newReplicasDidNotMatchError("a", "b")
+	tooManyClustersErr := newTooManyClustersError(10)
+	ingestionRateLimitedErr := newIngestionRateLimitedError(10, 10)
+
+	type testStruct struct {
+		err                error
+		expectedHTTPStatus int
+		expectedGRPCStatus codes.Code
+		expectedErrorMsg   string
+	}
+	testCases := map[string]testStruct{
+		"a generic error gets translated into a HTTP 503": {
+			err:                originalErr,
+			expectedHTTPStatus: http.StatusServiceUnavailable,
+			expectedGRPCStatus: codes.Internal,
+			expectedErrorMsg:   originalMsg,
+		},
+		"a DoNotLog of a generic error gets translated into a HTTP 503": {
+			err:                middleware.DoNotLogError{Err: originalErr},
+			expectedHTTPStatus: http.StatusServiceUnavailable,
+			expectedGRPCStatus: codes.Internal,
+			expectedErrorMsg:   originalMsg,
+		},
+		"a context.DeadlineExceeded gets translated into a HTTP 503": {
+			err:                context.DeadlineExceeded,
+			expectedHTTPStatus: http.StatusServiceUnavailable,
+			expectedGRPCStatus: codes.Internal,
+			expectedErrorMsg:   context.DeadlineExceeded.Error(),
+		},
+		"a replicasDidNotMatchError gets translated into an HTTP 202": {
+			err:                replicasNotMatchErr,
+			expectedHTTPStatus: http.StatusAccepted,
+			expectedGRPCStatus: codes.OK,
+			expectedErrorMsg:   replicasNotMatchErr.Error(),
+		},
+		"a DoNotLogError of a replicasDidNotMatchError gets translated into an HTTP 202": {
+			err:                middleware.DoNotLogError{Err: replicasNotMatchErr},
+			expectedHTTPStatus: http.StatusAccepted,
+			expectedGRPCStatus: codes.OK,
+			expectedErrorMsg:   replicasNotMatchErr.Error(),
+		},
+		"a tooManyClustersError gets translated into an HTTP 400": {
+			err:                tooManyClustersErr,
+			expectedHTTPStatus: http.StatusBadRequest,
+			expectedGRPCStatus: codes.InvalidArgument,
+			expectedErrorMsg:   tooManyClustersErr.Error(),
+		},
+		"a DoNotLogError of a tooManyClustersError gets translated into an HTTP 400": {
+			err:                middleware.DoNotLogError{Err: tooManyClustersErr},
+			expectedHTTPStatus: http.StatusBadRequest,
+			expectedGRPCStatus: codes.InvalidArgument,
+			expectedErrorMsg:   tooManyClustersErr.Error(),
+		},
+		"a validationError gets translated into an HTTP 400": {
+			err:                newValidationError(originalErr),
+			expectedHTTPStatus: http.StatusBadRequest,
+			expectedGRPCStatus: codes.InvalidArgument,
+			expectedErrorMsg:   originalMsg,
+		},
+		"a DoNotLogError of a validationError gets translated into an HTTP 400": {
+			err:                middleware.DoNotLogError{Err: newValidationError(originalErr)},
+			expectedHTTPStatus: http.StatusBadRequest,
+			expectedGRPCStatus: codes.InvalidArgument,
+			expectedErrorMsg:   originalMsg,
+		},
+		"an ingestionRateLimitedError gets translated into an HTTP 429": {
```

Comment:
IMO it makes sense to explicitly test that `ingestionRateLimitedError` gets translated into a `429` error both when `distributor.service-overload-status-code-on-rate-limit-enabled` is true and false. 
The other method (`distributor.toHTTPStatus()`) gives 2 different values when the cli flag is set to true and false.